### PR TITLE
IndexedDB: Import blink's tests for getAll/getAllKeys

### PR DESCRIPTION
**Pull request by inexorabletash**
Adds tests for new methods:

IDBObjectStore.getAll()
IDBObjectStore.getAllKeys()
IDBIndex.getAll()
IDBIndex.getAllKeys()

Present in Chrome Canary (likely to ship in M48), and experimental in FF41.
Specified in Editor's Draft: https://w3c.github.io/IndexedDB/

### CODE REVIEW

**Comment by inexorabletash** [comment]
Remind me to amend the commit comment for these. (Chromium work flow ignores the comments and uses what you put in the PR-equivalent, so I get lazy.)

File: IndexedDB/idbindex_getAll.html [hunk 1]
```diff
@@ -0,0 +1,230 @@
...
+      index = store.createIndex('test_idx', 'upper');
+    };
+  };
+}
+
+function createGetAllRequest(t, storeName, connection, range, maxCount) {
+    var transaction = connection.transaction(storeName, 'readonly');
+    var store = transaction.objectStore(storeName);
+    var index = store.index('test_idx');
+    // TODO(cmumford): Simplify once crbug.com/335871 is fixed.
```
Comment (Velmont) [inline IndexedDB/idbindex_getAll.html]: such stuff, and the workaround don't belong in spec tests :)

File: IndexedDB/idbindex_getAll.html [hunk 2]
```diff
@@ -0,0 +1,230 @@
...
+}
+
+doSetup(location.pathname + '-IDBIndex.getAll', 1, function(evt) {
+    var connection = evt.target.result;
+    async_test(function(t) {
+      var req = createGetAllRequest(t, 'out-of-line', connection, 'C');
+      req.onsuccess = t.step_func(function(evt) {
+          var data = evt.target.result;
+          assert_class_string(data, 'Array', 'result should be an array');
+          assert_array_equals(data.map(e => e.ch), ['c']);
```
Comment (Velmont) [inline IndexedDB/idbindex_getAll.html]: Can we use ES6?  I'd love that, but you don't use it elsewhere.

File: IndexedDB/idbobjectstore_getAll.html [hunk 3]
```diff
@@ -0,0 +1,174 @@
...
+          t.done();
+      });
+    }, 'Get all values');
+
+    async_test(function(t) {
+      var req = createGetAllRequest(t, 'out-of-line', connection, undefined,
+                                    10);
+      req.onsuccess = t.step_func(function(evt) {
+          assert_array_equals(evt.target.result,
+              ['a', 'b', 'c', 'd', 'e', 'f', 'g', 'h', 'i', 'j']
```
Comment (Velmont) [inline IndexedDB/idbobjectstore_getAll.html]: You don't "abcdef".split() anymore?  (I see no problem being explicit, just ask :P)

**Comment by Velmont** [comment]
The testing style is good here, -- there's a test for each actual test :)  Very good.

I'd love if we can start using ES6 for the tests.  Though that makes them fail for the wrong reasons in the end.


I think you should remove the workaround before merge though.  Feel free to merge when that is done.  Also, you want to commit amend (reminding), probably to prefix "IndexedDB:"? :)


Thanks for doing this btw!

File: IndexedDB/idbindex_getAll.html (hunk 2 above)
Comment (inexorabletash) [inline IndexedDB/idbindex_getAll.html]: These are upstreamed Blink tests where we could assume some ES6isms. It's not too uglifying to use function, so might as well.

**Comment by inexorabletash** [comment]
@Velmont - thanks for the feedback. Updated - please take another look?

**Comment by Velmont** [comment]
Goodie good :) It looks good, but it doesn't run for me!  It should fail, and after stepping through it, it seems to do so.  But it just keeps on hanging rather than reporting those failures.  I have no idea what the error might be, it's been such a long time since I had a look.

The exception is of course the getAllKeys missing, but that should be propegated up to the async_test() starter (and indeed is), and it should set itself to failed and do a test.done().  All this seems to happen correctly, but then it never gets anywhere.

Maybe @jgraham will instantly recognize what's been done wrong?

**Comment by inexorabletash** [comment]
huh, I'll give it a whirl in an old-fashioned browser. ;-) Likely there's a `t.step_func()` missing somewhere.

**Comment by inexorabletash** [comment]
Try this as the guts of a testharness page:
```js
setTimeout(foo, 0);
function foo() {
    alert('1');
    async_test(function(t) { throw Error(); });
    alert('2');
    async_test(function(t) { throw Error(); });
};
```
When the first alert fires the page displays 1/1 test failed. When OK is clicked, then clears the page and hangs with "Running, 1 complete, 1 remain". If `foo()` is called directly, not as part of a handler, it reports 2/2 fails as expected.

So... I think it's a testharness issue that a *synchronous* error in the body func of an `async_test` confuses the handler if the `async_test()` calls are *themselves* made asynchronously (IDB onsuccess, setTimeout, etc).

**Comment by inexorabletash** [comment]
@Velmont - added `{explicit_done:true}` which seems like the best approach here, verified the tests still pass in Chrome Canary and fail in downlevel browsers. Can you take a look and merge if it looks okay?

**Comment by Velmont** [comment]
> So... I think it's a testharness issue that a synchronous error in the body func of an async_test confuses the handler if the async_test() calls are themselves made asynchronously (IDB onsuccess, setTimeout, etc).

Yes, that was exactly what it looked like :)  Thanks for reproducing, investigating, coming up with reduced testcase and filing the bug!

The tests fails nicely on unsupported browsers now.  I have squashed and put this on master :)

Again, thanks a lot for doing this!

**Pull request closed without merging**